### PR TITLE
Reduce responsiveness impact by map canvas preview jobs

### PR DESCRIPTION
**Pull request by m-kuhn**
Starting 8 jobs in a loop on the main thread still can take some time (given that each job can have a larger number of layers).

This pull request starts each job separately in a quick succession but still with a small window in between for the event loop to process user input.

And it caches some database information that can lead to I/O during the setup of a layer source.

@mhugent @nyalldawson 
do you think this is a reasonable approach?

### CODE REVIEW

File: src/providers/postgres/qgspostgresprovider.cpp [hunk 1]
```diff
@@ -3950,9 +3950,21 @@ QgsCoordinateReferenceSystem QgsPostgresProvider::crs() const
...
+    else
+    {
+      QgsPostgresConn *conn = connectionRO();
+      conn->lock();
+      QgsPostgresResult result( conn->PQexec( QStringLiteral( "SELECT proj4text FROM spatial_ref_sys WHERE srid=%1" ).arg( srid ) ) );
+      conn->unlock();
+      if ( result.PQresultStatus() == PGRES_TUPLES_OK )
+      {
+        srs = QgsCoordinateReferenceSystem::fromProj4( result.PQgetvalue( 0, 0 ) );
+        sCrsCache.insert( srid, srs );
```
Comment (nyalldawson) [inline src/providers/postgres/qgspostgresprovider.cpp]: Better protect this with a mutex. This method may be called from non-main threads.

**Review by nyalldawson — COMMENTED** [review]
I like f33059d (when it's made thread safe).

I'll take your word that fc0b085 makes a difference. Honestly, I'm on the fence about reverting the preview job work anyway. I've had one crash since merging that (in the xyz provider - probably was unrelated, but the trace showed it happened during a preview job). I've also subjectively noticed that rendering seemed "more laggy" (again - may be unrelated.) At the very least I want to delay commencement of the preview jobs by at least 1 second after the main render is complete - that should help avoid any regressions due to this feature when rapidly redrawing the canvas.... (like i'll be doing during my presentation tomorrow ;) )

**Comment by nirvn** [comment]
I've also noticed UI feeling more laggy recently. And I've had a couple of random pan/zoom crashers in the last >1 week.

**Comment by m-kuhn** [comment]
There's noone with a backtrace by any chance?

My feeling is, that the preview isn't responsible for the crashes, it might just trigger them more often.

File: src/providers/postgres/qgspostgresprovider.cpp (hunk 1 above)
Comment (m-kuhn) [inline src/providers/postgres/qgspostgresprovider.cpp]: done

**Comment by m-kuhn** [comment]
> At the very least I want to delay commencement of the preview jobs by at least 1 second after the main render is complete

Hmmm... I somehow don't like this, that's just too long for layers that render quickly. Maybe make the waiting period dependant on the rendering time? And profile the render job preparation code. I remember e.g. creating QImages took some time, might be possible to cache them.

**Comment by nyalldawson** [comment]
> Hmmm... I somehow don't like this, that's just too long for layers that render quickly. 

Is it really a huge burden though if this doesn't kick in immediately? I doubt users would really ever notice the difference (as opposed to noticing the difference due to less responsiveness)

What about 0.5 seconds?

**Comment by m-kuhn** [comment]
Have you tried the latest version?

**Comment by nirvn** [comment]
@m-kuhn , @nyalldawson , on my machine, only one preview job is ever rendered (the top-left preview job), see this video:
[top-left-only-preview.zip](https://github.com/qgis/QGIS/files/1210253/top-left-only-preview.zip)

**Comment by m-kuhn** [comment]
Should be fixed by https://github.com/qgis/QGIS/pull/5006

**Pull request merged**